File: code-examples/pymongo/indexes/compound.py
from pymongo import MongoClient, ASCENDING

# Replace the following with your MongoDB deployment's connection string.
uri = "mongodb+srv://<user>:<password>@<cluster-url>?w=majority"
client = MongoClient(uri)

try:
    # Connect to the "sample_mflix" database
    database = client["sample_mflix"]
    # Access the database's "movies" collection
    movies = database["movies"]

    # Create an ascending index on the "type" and "genre" fields in the "movies" collection.
    result = movies.create_index([("type", ASCENDING), ("genre", ASCENDING)])
    print(f"Index created: {result}")

    # Define a query to find movies in the "Drama" genre
    query = {"type": "movie", "genre": "Drama"}
    # Define sorting criteria for the query results
    sort = [("type", ASCENDING), ("genre", ASCENDING)]
    # Include only the type and genre fields in the query results
    projection = {"_id": 0, "type": 1, "genre": 1}

    # Execute the query using the defined criteria and projection
    cursor = movies.find(query, projection).sort(sort)

except Exception as e:
    print(e)

finally:
    # Close the MongoDB client connection
    client.close()

File: code-examples/java-rs/indexes/compound.java
import com.mongodb.reactivestreams.client.MongoClients;
import com.mongodb.reactivestreams.client.MongoClient;
import com.mongodb.reactivestreams.client.MongoDatabase;
import com.mongodb.reactivestreams.client.MongoCollection;
import com.mongodb.reactivestreams.client.Success;
import org.bson.Document;

import static com.mongodb.client.model.Indexes.ascending;
import static com.mongodb.client.model.Filters.and;
import static com.mongodb.client.model.Filters.eq;
import static com.mongodb.client.model.Projections.excludeId;
import static com.mongodb.client.model.Projections.fields;
import static com.mongodb.client.model.Projections.include;
import static com.mongodb.client.model.Sorts.orderBy;

public class MainClass {

    public static void main(String[] args) {
        String uri = "mongodb+srv://<user>:<password>@<cluster-url>?writeConcern=majority";

        MongoClient mongoClient = MongoClients.create(uri);
        MongoDatabase database = mongoClient.getDatabase("sample_mflix");
        MongoCollection<Document> collection = database.getCollection("movies");

        // Creating Index
        collection.createIndex(ascending("type", "genre")).subscribe(success ->
                System.out.println("Index Created: " + success),
                throwable -> System.out.println("Failed to create index: " + throwable.getMessage()),
                () ->
                {
                    // Querying
                    collection.
                            find(and(eq("type", "movie"), eq("genre", "Drama")))
                            .projection(fields(include("type", "genre"), excludeId()))
                            .sort(orderBy(ascending("type"), ascending("genre")))
                            .forEach(document -> System.out.println(document.toJson()),
                                    throwable -> System.out.println("Query failed: " + throwable.getMessage()),
                                    () -> {
                                        // Closing the connection
                                        mongoClient.close();
                                    }
                            );
                }
        );
    }
}
